scipy/spatial/distance.py

```diff
@@ -1628,6 +1628,28 @@ def sokalsneath(u, v, w=None):
 _distance_wrap.pdist_correlation_double_wrap = _correlation_pdist_wrap
 _distance_wrap.cdist_correlation_double_wrap = _correlation_cdist_wrap
 
+def _distance_pybind_cdist_cosine(XA, XB, *, out=None, **kwargs):
+    # TODO: Remove the following code for weights
+    w = kwargs.pop('w', None)
+    if w is not None:
+        metric_info = _METRICS['cosine']
+        XA = np.ascontiguousarray(XA)
+        XB = np.ascontiguousarray(XB)
+        mA, n = XA.shape
+        mB, _ = XB.shape
+        XA, XB, typ, kwargs = _validate_cdist_input(
+            XA, XB, mA, mB, n, metric_info, **kwargs)
+        metric = metric_info.dist_func
+        return _cdist_callable(
+            XA, XB, metric=metric, out=out, w=w, **kwargs)
+
+    XA = XA.astype(np.float64)
+    XB = XB.astype(np.float64)
```

Comment:
If I comment these two lines out then the following test fail. Should I increase the tolerances (as the input is `np.float32` so minor increase the tolerances should be fine?)

```zshrc
========================================================== FAILURES ===========================================================
______________________________________ TestCdist.test_cdist_calling_conventions[cosine] _______________________________________
scipy/spatial/tests/test_distance.py:572: in test_cdist_calling_conventions
    self._check_calling_conventions(X1, X2, metric)
        X1         = array([[253.41678 , 133.03748 , 208.3706  , 226.47488 , 173.87625 ,
        290.92435 , 277.08063 , 260.64362 , 295.87...34 , 293.1113  , 128.6125  ,
        283.1046  , 275.94177 , 182.29692 , 294.8665  , 253.00142 ]],
      dtype=float32)
        X2         = array([[209.0965  , 144.3416  , 164.93024 , 167.71597 , 286.10556 ,
        213.18121 , 127.186516, 279.3654  , 244.08...34 , 287.3545  , 129.70238 ,
        291.80734 , 222.30965 , 134.87086 , 106.290665, 281.29544 ]],
      dtype=float32)
        eo_name    = 'random-float32-data'
        metric     = 'cosine'
        self       = <scipy.spatial.tests.test_distance.TestCdist object at 0x137676780>
scipy/spatial/tests/test_distance.py:551: in _check_calling_conventions
    assert_allclose(y1, y2, rtol=eps, verbose=verbose > 2)
        X1         = array([[253.41678 , 133.03748 , 208.3706  , 226.47488 , 173.87625 ,
        290.92435 , 277.08063 , 260.64362 , 295.87...34 , 293.1113  , 128.6125  ,
        283.1046  , 275.94177 , 182.29692 , 294.8665  , 253.00142 ]],
      dtype=float32)
        X2         = array([[209.0965  , 144.3416  , 164.93024 , 167.71597 , 286.10556 ,
        213.18121 , 127.186516, 279.3654  , 244.08...34 , 287.3545  , 129.70238 ,
        291.80734 , 222.30965 , 134.87086 , 106.290665, 281.29544 ]],
      dtype=float32)
        eps        = 1e-07
        kwargs     = {}
        metric     = 'cosine'
        self       = <scipy.spatial.tests.test_distance.TestCdist object at 0x137676780>
        y1         = array([[0.05484979, 0.08171444, 0.05579393, 0.05524987, 0.09750151,
        0.07073686, 0.08786177, 0.04963051, 0.0519..., 0.03029695, 0.07035159, 0.10272532, 0.06459693,
        0.06945175, 0.04259871, 0.06120552, 0.08087305, 0.10174133]])
        y2         = array([[0.05484983, 0.08171439, 0.05579386, 0.05524984, 0.0975015 ,
        0.07073683, 0.08786174, 0.04963045, 0.0519..., 0.03029696, 0.07035156, 0.10272536, 0.06459692,
        0.06945185, 0.04259878, 0.06120565, 0.08087306, 0.10174136]])
        y3         = array([[0.05484983, 0.08171439, 0.05579386, 0.05524984, 0.0975015 ,
        0.07073683, 0.08786174, 0.04963045, 0.0519..., 0.03029696, 0.07035156, 0.10272536, 0.06459692,
        0.06945185, 0.04259878, 0.06120565, 0.08087306, 0.10174136]])
/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/contextlib.py:81: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0
E   
E   Mismatched elements: 354 / 400 (88.5%)
E   Max absolute difference: 1.54077581e-07
E   Max relative difference: 5.8454892e-06
        args       = (<function assert_allclose.<locals>.compare at 0x14b80d1c0>, array([[0.05484979, 0.08171444, 0.05579393, 0.05524987, 0... 0.03029696, 0.07035156, 0.10272536, 0.06459692,
        0.06945185, 0.04259878, 0.06120565, 0.08087306, 0.10174136]]))
        func       = <function assert_array_compare at 0x110981760>
        kwds       = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=0', 'verbose': False}
        self       = <contextlib._GeneratorContextManager object at 0x11092a1e0>
_______________________________________ TestCdist.test_cdist_dtype_equivalence[cosine] ________________________________________
scipy/spatial/tests/test_distance.py:610: in test_cdist_dtype_equivalence
    assert_allclose(y1, y2, rtol=eps, verbose=verbose > 2)
        X1         = array([[ True, False,  True, False, False, False,  True, False, False,
         True, False,  True, False, False,  Tru...    [ True, False,  True,  True, False, False,  True,  True, False,
        False, False, False,  True, False,  True]])
        X2         = array([[ True,  True,  True,  True, False,  True, False, False,  True,
         True,  True,  True,  True,  True,  Tru...    [False,  True, False, False,  True, False,  True, False,  True,
         True,  True, False, False, False,  True]])
        eps        = 1e-07
        metric     = 'cosine'
        new_type   = <class 'numpy.float32'>
        self       = <scipy.spatial.tests.test_distance.TestCdist object at 0x1376a1610>
        test       = (array([[False,  True,  True, ..., False,  True,  True],
       [ True,  True,  True, ...,  True,  True,  True],
     ...ue, False, False]]), [<class 'numpy.uint64'>, <class 'numpy.int64'>, <class 'numpy.float32'>, <class 'numpy.float64'>])
        tests      = [(array([[False,  True,  True, ..., False,  True,  True],
       [ True,  True,  True, ...,  True,  True,  True],
    ... 135.6279  , 122.53102 , ..., 150.81123 , 231.02678 ,
        231.16676 ]], dtype=float32), [<class 'numpy.float64'>])]
        y1         = array([[0.41074435, 0.3827866 , 0.42264973, 0.3827866 , 0.45566895,
        0.22848325, 1.        , 0.5669873 , 0.7278..., 0.6619383 , 0.4023857 , 0.42857143, 0.42857143,
        0.53708995, 0.49290745, 0.78178211, 0.62203553, 0.71428571]])
        y2         = array([[0.41074436, 0.38278661, 0.42264974, 0.38278661, 0.45566897,
        0.22848326, 1.        , 0.56698731, 0.7278..., 0.66193829, 0.40238569, 0.4285714 , 0.4285714 ,
        0.53708995, 0.49290744, 0.7817821 , 0.62203552, 0.7142857 ]])
/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/contextlib.py:81: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0
E   
E   Mismatched elements: 6 / 400 (1.5%)
E   Max absolute difference: 4.74261168e-08
E   Max relative difference: 3.31982928e-07
        args       = (<function assert_allclose.<locals>.compare at 0x14b87f4c0>, array([[0.41074435, 0.3827866 , 0.42264973, 0.3827866 , 0... 0.66193829, 0.40238569, 0.4285714 , 0.4285714 ,
        0.53708995, 0.49290744, 0.7817821 , 0.62203552, 0.7142857 ]]))
        func       = <function assert_array_compare at 0x110981760>
        kwds       = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=0', 'verbose': False}
        self       = <contextlib._GeneratorContextManager object at 0x11092a1e0>
=================================================== short test summary info ===================================================
FAILED scipy/spatial/tests/test_distance.py::TestCdist::test_cdist_calling_conventions[cosine] - AssertionError: 
FAILED scipy/spatial/tests/test_distance.py::TestCdist::test_cdist_dtype_equivalence[cosine] - AssertionError: 
================================== 2 failed, 1517 passed, 7 skipped, 20 deselected in 8.94s ===================================
```